tls/extensions/s2n_client_supported_groups.c

```diff
@@ -58,9 +58,38 @@ int s2n_recv_client_supported_groups(struct s2n_connection *conn, struct s2n_stu
     proposed_curves.data = s2n_stuffer_raw_read(extension, proposed_curves.size);
     notnull_check(proposed_curves.data);
 
-    if (s2n_ecc_evp_find_supported_curve(&proposed_curves, &conn->secure.server_ecc_evp_params.negotiated_curve) != 0) {
-        /* Can't agree on a curve, ECC is not allowed. Return success to proceed with the handshake. */
-        conn->secure.server_ecc_evp_params.negotiated_curve = NULL;
+    s2n_parse_client_supported_groups_list(&proposed_curves, conn->secure.mutually_supported_groups);
+    s2n_choose_supported_group(&conn->secure.server_ecc_evp_params, conn->secure.mutually_supported_groups);
+    return 0;
+}
+
+int s2n_parse_client_supported_groups_list(struct s2n_blob *iana_ids, struct s2n_ecc_evp_params *supported_groups) {
+    struct s2n_stuffer iana_ids_in = {0};
+
+    GUARD(s2n_stuffer_init(&iana_ids_in, iana_ids));
+    iana_ids->data = s2n_stuffer_raw_write(&iana_ids_in, iana_ids->size);
+    
+    for (int i = 0; i < iana_ids->size / 2; i++) {
+        uint16_t iana_id;
+        GUARD(s2n_stuffer_read_uint16(&iana_ids_in, &iana_id));
+        for (int j = 0; j < s2n_ecc_evp_supported_curves_list_len; j++) {
+            const struct s2n_ecc_named_curve *supported_curve = s2n_ecc_evp_supported_curves_list[j];
+            if (supported_curve->iana_id == iana_id) {
+                supported_groups[j].negotiated_curve = supported_curve;
+            }
+        }
+    }
+    return 0;
+}
+
+int s2n_choose_supported_group(struct s2n_ecc_evp_params *chosen_group, struct s2n_ecc_evp_params *group_options) {
```

Comment:
do we have the assumption that group_options's length == s2n_ecc_evp_supported_curves_list_len? would be useful to add this check?

Reply:
I can add a check, that's pretty easy.